pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,111 @@
+//go:build windows
+
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	core "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/controller"
+	"github.com/openshift/windows-machine-config-operator/pkg/daemon/manager"
+	"github.com/openshift/windows-machine-config-operator/pkg/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/pkg/servicescm"
+)
+
+// Deconfigure removes an instance from the cluster. If preserveNode is true, the node object is not deleted.
+// If we are able to get the services ConfigMap tied to the desired version, all services defined in it are cleaned up.
+// TODO: Otherwise, perform cleanup based on a combination of the OpenShift managed tag and latest services ConfigMap.
+func Deconfigure(cfg *rest.Config, ctx context.Context, preserveNode bool) error {
+	// Cannot use a cached client as no manager will be started to populate cache
+	directClient, err := controller.NewDirectClient(cfg)
+	if err != nil {
+		return errors.Wrap(err, "could not create authenticated client from service account")
+	}
+
+	var node *core.Node
+	cm := &core.ConfigMap{}
+	// Fetch the CM of the desired version
+	err = func() error {
+		addrs, err := controller.LocalInterfaceAddresses()
+		if err != nil {
+			return err
+		}
+		node, err = controller.GetAssociatedNode(directClient, addrs)
+		if err != nil {
+			return err
+		}
+		desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+		if !present {
+			return errors.Wrapf(err, "node missing desired version annotation")
+		}
+		return directClient.Get(ctx,
+			client.ObjectKey{Namespace: controller.WMCONamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	}()
```

Comment:
The logic deciding if a full (with correct configmap) cleanup should occur is no longer being unit tested, it would be nice if it was considering we have the capabilities to.

moving this logic into a new function `func normalCleanupPossible (bool,error)` would
1) make the code more readable
2) Enable us to make the distiniction between an actual error with interacting with the API, versus the required node annotation/config being missing, and handle that appropriately.
3) Unit test this logic

Reply:
feel free to change the name, this is just an example

Reply:
@aravindhp any thoughts on this?
after discussing with @saifshaikh48 this has proven more difficult than i originally thought
specifically having to consider all the cases of what should be an error that causes program failure and what should cause the best effort cleanup

Reply:
I've suggested at the least doing
```
	addrs, err := controller.LocalInterfaceAddresses()
	if err != nil {
		return err
	}
	var node *core.Node
	var cmData *servicescm.Data
	err := func() error {
		node, err = controller.GetAssociatedNode(directClient, addrs)
		if err != nil {
			return err
		}
		desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
		if !present {
			return errors.Wrapf(err, "node missing desired version annotation")
		}
		cm := &core.ConfigMap{}
		err = directClient.Get(ctx,
			client.ObjectKey{Namespace: controller.WMCONamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
		if err != nil {
			return err
		}
		cmData, err = servicescm.Parse(cm.Data)
		if err != nil {
			return err
		}
	}()
```
But I feel like we are missing out on the case that a `Node list`, or `ConfigMap Get` fails due to whatever reason, and WMCO re-reconciling the cleanup would fix things, but the cleanup chooses instead to resort to best effort cleanup.

This may not be important enough to hold up this PR but it might be good to have a TODO?
Or am I overthinking things here?

Reply:
We have a similar problem if getting one of the services fail. But what I don't understand is why resorting to best effort clean is bad. As long as we don't delete the node object on intermittent failure, we should be fine with WMCO re-reconciling on failures. Or am I missing something?

Reply:
I'm probably overthinking things then
the recently pushed changes are fine with me

Reply:
Yep, there should be no issue if WMCO re-reconciles on failure. 

I've incorporated the above suggested code.